src/common/tests_dmg_helpers.c

```diff
@@ -750,6 +750,123 @@ dmg_pool_destroy(const char *dmg_config_file, const uuid_t uuid, const char *grp
 	return rc;
 }
 
+static int
+dmg_pool_target(const char *cmd, const char *dmg_config_file, const uuid_t uuid,
+		const char *grp, d_rank_t rank, int tgt_idx)
+{
+	char			uuid_str[DAOS_UUID_STR_SIZE];
+	int			argcount = 0;
+	char			**args = NULL;
+	struct json_object	*dmg_out = NULL;
+	int			rc = 0;
+
+	uuid_unparse_lower(uuid, uuid_str);
+	args = cmd_push_arg(args, &argcount, "%s ", uuid_str);
+	if (args == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	if (grp != NULL) {
+		args = cmd_push_arg(args, &argcount, "--sys=%s ", grp);
+		if (args == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+	}
+
+	if (tgt_idx >= 0) {
+		args = cmd_push_arg(args, &argcount, "--target-idx=%d ", tgt_idx);
+		if (args == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+	}
+
+	args = cmd_push_arg(args, &argcount, "--rank=%d ", rank);
+	if (args == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	rc = daos_dmg_json_pipe(cmd, dmg_config_file,
+				args, argcount, &dmg_out);
+	if (rc != 0) {
+		D_ERROR("dmg failed");
+		goto out_json;
+	}
+
+out_json:
+	if (dmg_out != NULL)
+		json_object_put(dmg_out);
+	cmd_free_args(args, argcount);
+out:
+	return rc;
+}
+
+int
+dmg_pool_exclude(const char *dmg_config_file, const uuid_t uuid,
+		 const char *grp, d_rank_t rank, int tgt_idx)
+{
+	return dmg_pool_target("pool exclude", dmg_config_file, uuid, grp, rank, tgt_idx);
+}
+
+int
+dmg_pool_reintegrate(const char *dmg_config_file, const uuid_t uuid,
+		     const char *grp, d_rank_t rank, int tgt_idx)
+{
+	return dmg_pool_target("pool reintegrate", dmg_config_file, uuid, grp, rank, tgt_idx);
+}
+
+int
+dmg_pool_drain(const char *dmg_config_file, const uuid_t uuid,
+	       const char *grp, d_rank_t rank, int tgt_idx)
+{
+	return dmg_pool_target("pool drain", dmg_config_file, uuid, grp, rank, tgt_idx);
+}
+
+int
+dmg_pool_extend(const char *dmg_config_file, const uuid_t uuid,
+		const char *grp, d_rank_t *ranks, int rank_nr)
+{
+	char			uuid_str[DAOS_UUID_STR_SIZE];
+	d_rank_list_t		rank_list = { 0 };
+	char			*rank_str = NULL;
+	int			argcount = 0;
+	char			**args = NULL;
+	struct json_object	*dmg_out = NULL;
+	int			rc = 0;
+
+	uuid_unparse_lower(uuid, uuid_str);
+	args = cmd_push_arg(args, &argcount, "%s ", uuid_str);
+	if (args == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	if (grp != NULL) {
+		args = cmd_push_arg(args, &argcount, "--sys=%s ", grp);
+		if (args == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+	}
+
+	rank_list.rl_ranks = ranks;
+	rank_list.rl_nr = rank_nr;
+
+	rank_str = d_rank_list_to_str(&rank_list);
+	if (rank_str == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
```

Comment:
if rank_str == NULL here, then args will be not NULL and when the code reaches the out label the args will not be freed. So before jumping to the out label perhaps cmd_free_args() could be called here?

Reply:
or simply goto the out_json label ?

Reply:
Oops, good catch. I moved things around a little; hopefully it makes sense now. I guess another option would be to just get rid of the `out_rankstr:` label and put the `D_FREE(rank_str)` under `out:`? It's safe to call that even on a null pointer, I think. Maybe more sloppy that way though.